### ISSUE TITLE
Property declarations can silently mask DSL methods

### ISSUE THREAD

**Issue by lamont-granquist** [chef/chef#4006]
With new-style converged resources in 12.5.0:

``` ruby
property :name, String, name_property: true
property :template, String  # <--- this is badly named

action :whatever do
  template "/tmp/#{name}" do
    source "#{name}.erb"
  end
end
```

That is just `new_resource.template = "/tmp/#{name}` with a block that is silently ignored.  It never creates a resource.

Because we dropped the need to do `new_resource.whatever` or `params[:whatever]` and the properties are methods they conflict with the DSL resources which are also methods.

The way that it works is probably fine, and declare_resource or renaming the property works around the issue, but I lost 2 hours debugging this and we should probably try to warn.

**Comment by coderanger** [comment]
The solution I went with in `Poise::Fused` is that inside `action` you get the actual provider context, not the resource. That is kind of gross though :-/

**Comment by lamont-granquist** [comment]
But, I suspect that being able to do `my_property` instead of `new_resource.my_property` is a design goal of the new style resources, @jkeiser?

**Comment by lamont-granquist** [comment]
As far as warning goes, though, we're going to have an issue there as well.  If we're warning when properties are defined, we're still parsing through LWRPs and we don't know if a property will overwrite some resource which is later defined.  We could at least reliably catch properites overwriting core chef resources, which would still be useful.

We could do some kind of horrible hook after parsing LWRPs which walked them all and checked all their properties against the DSL methods and warned about any namespace overlap, but that is mildly awful.

If we track everything we add to the DSL we could check every time we assign to a property, but that is a bit more overhead to do that every time, and it'll emit a warning every time the bad property is used -- we could emit only if a property is passed a block to discover attempted use of a property-as-a-resource, but not all resource declarations have a block so that misses `package "foo"` when you've got package-as-a-property.

**Comment by jkeiser** [comment]
I believe we can make it warn--make it so that when you are declaring an inline resource, if you *read* a property, it warns if it's a dup of the enclosing provider.

**Comment by jkeiser** [comment]
Should warn with https://github.com/chef/chef/pull/4269

**Comment by jkeiser** [comment]
@lamont-granquist oh! There is a way to fix this one that we should really do anyway. If you pass a block to a property and it doesn't need one, we should raise an error.

**Comment by lamont-granquist** [comment]
Yes, that sounds better than all the other approaches I was thinking of.

**Comment by jkeiser** [comment]
The sad bit is there would be no way to actually use a template; while I still like the implicit methods, I think we're going to need a separate name like `recipe` so you *can* disambiguate when you want (for example, in this case you could say `recipe.template ...`).

**Comment by coderanger** [comment]
Cramming the resource, provider, and a recipe DSLs into one namespace is going to keep having these kinds of issues :-/

**Comment by lamont-granquist** [comment]
Yes, and since you have to use new_resource.foo _sometimes_ its probably going to be my best practice recommendation to use it _all the time_.  Then users can just write code without becoming experts in thinking about what context they are in.

**Comment by coderanger** [comment]
That was my conclusion too, but I agree it is weird and annoying.

**Comment by jkeiser** [comment]
I actually think it would be a lot less weird and annoying if it was not named `new_resource`. Like, `desired`.  `desired.template`, `desired.file`, etc.

**Comment by lamont-granquist** [comment]
this was fixed in a breaking change in 13.0.x
